addons/addon-base-raas/packages/base-raas-services/lib/data-source/access-strategy/roles-only/filesystem-role-service.js

```diff
@@ -0,0 +1,553 @@
+    //   - Add usage of the study by this role, see item #1 above.
+
+    // Get app role entity
+    const appRoleEntity = await appRoleService.mustFind(requestContext, { accountId, bucket, arn: appRoleArn });
+
+    // Get all existing usages by fs roles for this study (keeping env permission in mind)
+    const fsRoleUsages = await resourceUsageService.getResourceUsage(requestContext, {
+      resource: studyResourceId,
+      setName: fsUsageSetName,
+    });
+    const fsRoleEntities = _.get(fsRoleUsages, fsUsageSetName, []);
+
+    // Start a loop to see if existing fs roles for the study have the member account already
+    // or can take it without exceeding the trust doc limit?
+    let fsRoleEntity;
+
+    // eslint-disable-next-line no-restricted-syntax
+    for (const fsRoleArn of fsRoleEntities) {
+      const entity = await this.find(requestContext, { accountId, bucket, arn: fsRoleArn });
+      // eslint-disable-next-line no-continue
+      if (!entity) continue;
+
+      // Check if the provided study matches any of the studies in the fs role entity, including the env permissions
+      // eslint-disable-next-line no-continue
+      if (!hasStudy(entity, studyEntity)) continue;
+
+      if (hasMemberAccount(entity, memberAccountId)) {
+        // We found an existing fs role entity that already has the member account
+        fsRoleEntity = entity;
+
+        // Add environment usage
+        await resourceUsageService.addUsage(requestContext, {
+          resource: studyResourceId,
+          setName: memberAccountId,
+          item: envId,
+        });
+
+        break;
+      }
+
+      if (!maxReached(addMemberAccount(entity, memberAccountId))) {
+        // We found an existing fs role entity and we can fit the member account in its trust doc
+
+        // Update the fs role policy in the data source account
+        await this.updateAssumeRolePolicy(entity);
+
+        // Add environment usage
+        await resourceUsageService.addUsage(requestContext, {
+          resource: studyResourceId,
+          setName: memberAccountId,
+          item: envId,
+        });
+
+        // We need to save the fsRoleEntity (even though it is already in the database), because
+        // we added the member account id to the trust doc.
+        fsRoleEntity = await this.saveEntity(requestContext, entity);
+        break;
+      }
+    }
+
+    // We have one, we are done and we can return it
+    if (fsRoleEntity) return fsRoleEntity;
+
+    // Create a new fs role entity
+    fsRoleEntity = newFsRoleEntity(appRoleEntity);
+    addStudy(fsRoleEntity, studyEntity);
+    addMemberAccount(fsRoleEntity, memberAccountId);
+
+    // Create in the fs role in the data source account
+    await this.provisionRole(fsRoleEntity);
+
+    // Add environment usage
+    await resourceUsageService.addUsage(requestContext, {
+      resource: studyResourceId,
+      setName: memberAccountId,
+      item: envId,
+    });
```

Comment:
nit: Does adding environment usage really belong here given that this method and file is just about file system roles.

Reply:
This is actually core to the filesystem **allocateRole** logic.  The logic is a 'pool manager' logic (checkin/checkout), it needs to keep track of what roles were issued and do a clean up after that.  If we move this logic outside of the filesystem role service, we need another service called 'filesystem role pool service'.   When I thought about doing that, I realized that we can combine them and there is no need for an additional service.  Let me know what you think.